### ISSUE TITLE
FTP File Upload on Windows - Data Socket Sometimes isnt Closed

### ISSUE THREAD

**Issue by Snork2** [bagder/curl#485]
I've got a problem with libcurl 7.44.0 on Windows 7 x64

Sometimes under heavy CPU load and with multiple `curl_easy_perform` ftp uploading threads, libcurl fails to close FTP data socket.

In file `lib/ftp.c` (my comments start with //!!)

    //!! line 3324
    #ifdef _WIN32_WCE /
      shutdown(conn->sock[SECONDARYSOCKET], 2);  /* SD_BOTH */   //!! please note this ifdef with _WIN32_WCE, is it supposed to be so?
    #endif

    <...>

    //!! line 3348
        if(CURL_SOCKET_BAD != conn->sock[SECONDARYSOCKET]) {
          Curl_closesocket(conn, conn->sock[SECONDARYSOCKET]);      //!! here it is, this operation silently fails to close socket sometimes
          conn->sock[SECONDARYSOCKET] = CURL_SOCKET_BAD;
          conn->bits.tcpconnect[SECONDARYSOCKET] = FALSE;
        }

Then, libcurl just hangs in a `result = Curl_GetFTPResponse(&nread, conn, &ftpcode);` (line 3336) operation, until data socket is being killed with TCP RST signal (I believe operating system do this, since I saw RSTs from server and from client). Such situation is interpreted by FTP server as an error, it reports it through control socket and then libcurl reports error with:

    //!! line 3389
    if(!ftpc->dont_check) {
      /* 226 Transfer complete, 250 Requested file action okay, completed. */
      if((ftpcode != 226) && (ftpcode != 250)) {
        failf(data, "server did not report OK, got %d", ftpcode);
        result = CURLE_PARTIAL_FILE;
      }
    }

I believe, though not sure, that noted `#ifdef _WIN32_WCE` block with `shutdown(conn->sock[SECONDARYSOCKET], 2);  /* SD_BOTH */` code is an error.
When I’ve changed ifdef to `#ifdef _WIN32` my problem has gone. Socket is being closed correctly every time.

My setup:

- Windows 7 x64
- msvc 9 2008 (x86 target arch)
- libcurl 7.44.0 (libcurl-vc9-x86-debug-dll-ssl-dll-zlib-dll-ssh2-dll-ipv6-sspi)
- openssl 1.0.2d
- libssh2 libssh2 1.6.0
- zlib 1.2.8
- Tested with Windows Server 2008 IIS FTP Server and FileZilla Server 0.9.53 beta

**Comment by bagder** [comment]
So how can the socket close sometimes "fail" ? That doesn't make any sense to me. A socket should not have to be shutdown before it gets closed (the ifdef there is only present because we know windows CE to be crazy) and I also don't understand what server load or multiple transfers have to do with it?

Curl_GetFTPResponse() waits for a response on the control connection. Presumably then it doesn't get a response until the successful close of the data connection?

I'm mostly just afraid that reversing the logic and doing shutdown() unconditionally for all windows builds will trigger some other unexpected side-effects.

**Comment by Snork2** [comment]
> So how can the socket close sometimes "fail" ? That doesn't make any sense to me. A socket should not have to be shutdown before it gets closed (the ifdef there is only present because we know windows CE to be crazy) and I also don't understand what server load or multiple transfers have to do with it?

By failure I meant, that data connection, as seen by Process Explorer, remained in `ESTABLISHED` state, and there was no FIN TCP packets on sniffer after `Curl_closesocket(conn, conn->sock[SECONDARYSOCKET]);`. About server load and multiple transfers - it is conditions under which I observed an erroneous behavior, with single thread everything was fine. I tried to provide maximum information to reproduce the error, but in the end I didn't note the most crucial condition which I think now - was the reason of error I've encountered.

My program, which is using liburl, spawns child processes and redirects their stdout and stderr. For that, it uses INHERIT_HANDLES flag in WINAPI `CreateProcess` call. It seems that socket handles, used by libcurl somehow was inherited by one of child processes and this is caused data socket to remain opened after lbcurl `closesocket` call. Shutdown initializes FIN packets exchange, FTP server then closes its socket, sends libcurl a success reply and the transfer considered successful by both ends. This is how I understand situation now.

> Curl_GetFTPResponse() waits for a response on the control connection. Presumably then it doesn't get a response until the successful close of the data connection?

Sure it is. FTP server waits for data connection closure to report success in reply to STORE command, but connection were not closed. On other side - libcurl waits for FTP response. Thus deadlock happens, which is resolved by OS, which kills idle data connection with TCP RST packet. This is what I had seen in debugger and sniffer.

Sorry, for false issue, I thought that every handle is not inheritable by default, but currently I observe even `std::ofstream` file handles being inherited.

Looks like its not a libcurl error. Should I close this issue?
